### DISCUSSION TITLE
CP2K Build System

### DISCUSSION THREAD

**Discussion by juerghutter** [cp2k/cp2k#1989] [General]
Building CP2K is one of the major problems of the user community. We should have an ongoing discussion on the
way we want to support users and developers in building and maintaining CP2K on all platforms.
Here is a link to a new system that joins Spack, EasyBuild, Singularity/Docker, etc.

Sustainable packaging of quantum chemistry software with the Nix package manager
Markus Kowalewski, Phillip Seeber, Int J Quantum Chem.2022;e26872.
https://doi.org/10.1002/qua.26872

**Comment by hfp** [comment]
We could check _why_ user's wanted to build CP2K on their own (user != person used to build complex software). I can understand that containers making a build more reproducible (pinning the OS) are not exactly what people want in some cases. Hence we could check if we can build downloadable binaries for more cases/flavors, e.g., statically linked, moderate GLIBC dependencies, with and without different GPU flavors, nightly builds as well as current release, etc.  Related to this there are also engineering tasks we can take on, e.g., making GPUs a runtime decision, building generic binaries with performance critical parts taken based on CPUID, etc.

**Comment by oschuett** [comment]
Thanks for sharing!

Yes, installing CP2K has indeed become quite challenging. The main reason are the many libraries that were added in recent years. So one obvious countermeasure would be to start pushing back on new libraries, especially when they pull in even more dependencies on their own.

The idea of distributions is to share the packaging work. However, most of our dependencies are very specialized, which means we have nobody to share the work with. Furthermore, there are many competing distributions - our [installation page](https://www.cp2k.org/download#from_a_distribution) currently list seven: _Debian_, _Fedora_, _Arch_, _Homebrew_, _EasyBuild_, _Spack_, and _Conda_. And according to that article we should apparently also add _Nix_. So I don't think we can hope to keep up. It should be the distributions' responsibility to find maintainers for their packages.

We as CP2K developers should instead focus on the [toolchain](https://github.com/cp2k/cp2k/tree/master/tools/toolchain), which serves three purposes:

1. It acts as a reference for the various package maintainers.
2. It enables our fast pace of development because changes don't have to be coordinated with others.
3. It builds our container images, which I believe will long-term replace most distributions anyways.

So my plan for the future is to keep investing into the toolchain, e.g. by migrating it to Python.

**Comment by juerghutter** [comment]
A probably incomplete list of dependencies:

System
+ Compilers (versions) : GCC, Intel, CCE
+ MPI : mpich, openmpi, intelmpi
        MPI_MODE ?
+ math-libraries : acml, cray, mkl, openblas
+ GPU : CUDA/NVCC, HIP, ROCm, OpenMP
        optimized kernels
+ cmake

Mandatory
+ lapack
+ scalapack
+ dbcsr (submodule)

Performance
+ libsmm / libxsmm
+ elpa
+ cosma
+ spla
+ spfft
+ fftw

Feature
+ libint
+ libxc
+ plumed
+ pexsi -> ptscotch, superlu
+ sirius -> hdf5, spfft, spla, cosma, libvdwxc, spglib, gsl
+ spglib
+ libvori
+ quip

External
+ i-Pi
+ ASE
+ Gromacs
+ MonaLisa
+ GRRM
+ PyRetis
+ AiiDA

**Comment by hfp** [comment]
If someone found it reasonable to look at binary packages, I believe a script can extract binaries from Docker images and build a classic archive. From the above list, most things listed as feature can be taken without side-effect and some of the cases can be ignored like CCE (and it would be practical to just focus on GCC). A counter-example of "all features can be combined" is MKL's FFTW3 interface not sufficient for some of the Sirius dependencies. Anyway, perhaps it helps also to only aim at nightlies along with a history of previous nightly versions. I think MPI flavors are important and a constant pain-point for users (Intel MPI, OpenMPI). After all, it's not too much (if automated) even if x86-64 and ARM64 would be another factor of two on top.

**Comment by juerghutter** [comment]
Just as a side note: The Feb. 22 new stable release of Gromacs includes the CP2K QM/MM module.
We can expect many new users with installation problems.

**Comment by oschuett** [comment]
> A probably incomplete list of dependencies:

I'd say it's pretty complete. Libint technically also depends on C++ Boost, but we hide it via a [workaround](https://github.com/cp2k/libint-cp2k). And QUIP pulls in the FoX xml library. I'm also trying to maintain a list of our dependencies [here](https://github.com/cp2k/cp2k/blob/master/tools/toolchain/README.md#licenses).

If CP2K continues to grow we might someday be able to vertically integrate some of our vital dependencies, e.g. by including our own libint or eigensolver. However, I'm afraid for now we're stuck with them.

> Hence we could check if we can build downloadable binaries for more cases/flavors, e.g., statically linked

Containers and static binaries are very similar in that they have a lot of hardware assumptions backed in. Due to combinatorics (`cpu_arch` x `gpu_arch` x `mpi_library` x ...) we'd end up with O(100) different configurations. With any luck, the hardware might become more standardized in the future, but current trends don't suggest that.

Maybe we could offer a build service that produces static binaries and container images on demand 🤔?

The other way forward that I can see is to make the Toolchain more portable. We already have [install_requirements_ubuntu.sh](https://github.com/cp2k/cp2k/blob/master/tools/toolchain/install_requirements_ubuntu.sh) and [install_requirements_fedora.sh](https://github.com/cp2k/cp2k/blob/master/tools/toolchain/install_requirements_fedora.sh). So we could also add e.g. `install_requirements_spack.sh` and `install_requirements_conda.sh`.

**Reply by rathann** [reply]
Note that Fedora has most of the dependencies packaged already and it would be best to re-use them instead of building them from scratch.

**Reply by oschuett** [reply]
Yes, I believe several users do this already and we actually [test it](https://github.com/cp2k/cp2k/blob/master/tools/docker/Dockerfile.test_gcc10#L41-L48) for Ubuntu. I guess, we could do the same for Fedora as well. Currently, we only [test the full toolchain](https://dashboard.cp2k.org/archive/fedora-psmp/index.html) on Fedora.

Btw, one recent addition that enabled many new use-cases is [SIRIUS](https://github.com/electronic-structure/SIRIUS). Do you think this would be feasible to add?

In any case, thank you very much for your great work!

**Comment by dev-zero** [comment]
From my point of view we should take care to not mix different concepts of build systems and make sure we support the minimal requirements first:

1. build supported configurations of CP2K in a quick and reliable way (make, ninja)
2. verify mandatory or optional requirements and dependencies  (cmake, autotools, etc.) ... basically codify and verify our requirements to ensure consistent builds and correct binaries (both unfortunately still an issue), giving the user meaningful feedback to fix any issues wrt dependencies. Together with automated discovery of already available libraries (no matter how they have been made available).

Any package manager/install scripts on top of that minimal feature set will mean additional complexity (for both developers and users) and only works if it supports the users use case:

- Spack/EasyBuild are exhaustive when it comes to installing extra dependencies. Example: if you're not telling them to reuse installed Perl, Python, etc. you end up installing much more than the direct dependencies of CP2K. Nix has a similar issue. Finding and reusing already installed dependencies can unfortunately be difficult.
- AFAIK, the relocation feature of Nix for example requires either an initial setup by the superuser, [user namespace support](https://github.com/nix-community/nix-user-chroot) (which some sites do not enable) or [may have performance impact](https://github.com/DavHau/nix-portable).
- Singularity, resp. [Apptainer](https://apptainer.org/) requires some system setup
- Root-less containers require some system setup (enabling of user namespaces)
- Extracting binaries from a container may or may not work, or something in-between. For static binaries it might work. But given that widespread "container vacuuming" (removing from a container what is not needed) but we rather have to rely on manual build stages suggests that this is not a trivial problem, even when staying in the confined space of containers.

(Almost) static binaries do not provide a solution from my point-of-view: dynamically linked device libraries (`libcuda`), system runtime libraries (MPI, BLAS, etc.) make that impossible, even with projects like [MPItrampoline](https://github.com/eschnett/MPItrampoline). Maybe for distributions like conda, a binary with trampolines for MPI, BLAS, LAPACK might be a good enough solution.

Containers on the other hand currently provide a solution for distribution and archival (partially solving the problem where a site gets an updated software stack), but unless we're saying "we only support CP2K executables built with this specific container configuration" will not solve step 1 and 2 from above. Even worse, any update of the container requires someone to manually verify correctness as they might change on what they install (so, step 2 must be solved anyway). Also, wider adoption might take at least one more generation of supercomputers. And containers are usually not optimized.

Wrt our own toolchain: every extension of it will move it more towards something like Spack, EasyBuild, Nix, Gentoo/Prefix, etc. Essentially replicating their efforts. From that point-of-view it would make more sense to provide a Spack environment configuration (and the EasyBuild or Nix correspondent configurations) and [provide a small wrapper around it](https://github.com/cp2k/cp2k/blob/master/tools/toolchain-spack/install_cp2k_toolchain.sh). But that would require people getting familiar with one or more solutions out there which takes more effort at first than starting (almost) from scratch and thus seems less rewarding.

Wrt providing on-demand containers: Docker (the company) struggled with that and had to severely limit pulls and rebuilds, the Singularity registry struggled with just the traffic cost, [GCP just raised their prices](https://news.ycombinator.com/item?id=30671997), hence unlikely an option.

And since Nix is mentioned, [GNU Guix](https://guix.gnu.org) might also be a candidate ;-)

**Reply by haampie** [reply]
> Finding and reusing already installed dependencies can unfortunately be difficult.

Note that: `spack external find python perl cmake` etc works decently in Spack 0.17.x and are picked up by default (does not require you to set --not-buildable or so), and concretization with `--reuse` will optimize for reuse of spack-installed packages (so you get older dependencies, but fewer builds, satisfying the package constraints). So I think this problem is taken seriously.

**Comment by mtaillefumier** [comment]
Building CP2K has become more challenging because the number of dependency increased substantially. This was pointed out by many persons in this thread and elsewhere. as @dev-zero pointed out we should not mix compiling a software, handling its dependencies,. package management,and software distribution. 

We have to separate at least three points in (order of importance) :
- CP2K configuration and compilation (configure cp2k with the dependencies available on the system)
- Dependency handling and package management (spack, easybuild etc...)
- testing, deployment, and software distribution. 

These three steps use different tools, and care should be taken to maintain separation of concerns (e.g. configuration (CMake/meson) is agnostic to the method for dependency handling (spack/easybuild)). A normal user should be able to do these three steps, the system been a workstation or a HPC cluster. Normal users may not be concerned about deployment and software distribution though as it  involves IT administration to some extend.

Compilation and configuration should rely on small set of standard tools that are designed for that purpose only and can be found on any system. It can be cmake, meson or any other standard tool with a large community. Using CMake for instance (or meson) is a reasonable choice as it is well integrated in spack, easybuild or any other major package manager and can be found on any HPC system or workstation. Separating dependencies based on their importance makes it easier to configure, build and test cp2k as well.
This step is probably the most important. It avoids miss-configurations, saves time as we do not have to tinker makefiles by hand every time we add a new dependency, new hardware support, or different programming paradigm (oneapi, cuda, hip to cite few of them). It also helps users, devs, hpc support groups  when  issues occur. 

A full DIY solution for building dependencies is not reasonable anymore as the number of dependency exploded in the past few years. It is source of many problems, CIs breaking, PR testing failing, etc. More importantly it does not scale as the custom solution would have to build the dependencies of the dependencies. Compiling/configuration  tools such as cmake should not be responsible for building cp2k dependencies either as it is a completely different set of problems.  Major HPC systems  use spack or easybuild while distributions will have their own solution based most of the time on standard tools such as rpm / apt / tarball. It is not a simple problem to solve  despite the use of standard tools and many engineers around the world keep working on these issues, so why not using what they have done already.

The two first points also simplify the creation of container for  testing, software distribution and deployment. Containers are good tools for many things (CI, etc...) but they should not be confused with a package management or a build system. As pointed out here distributed containers (or packages) have many drawbacks which means that we have to build cp2k by hand or use a package manager configured for many platforms to effectively install it. So while we can provide the recipe for building a generic container for commodity hardware on which cp2k can run I do not think that distributing the container itself is practical. Many users will just rely on their distribution package manager for that and devs / HPC specialists will compile cp2k by hand or tools such as spack or easybuild. 
Containers are very useful though. CI is one area that benefit from it, automatic testing and hpc deployment (although it is a package management issue) an other one. 

Finally normal users should be able to validate the correctness of a compilation without the need to build a container (building a container and running is not possible by default on many distributions). Ideally a simple make check should do. The regtests suite provides this functionality and can easily be modified to take a conventional build system into account.

**Comment by hfp** [comment]
I agree with Ole about continuing the Toolchain (still a reasonable Spack recipe may be desired). I believe it is possible to leverage the Toolchain more and perhaps solve some recurring issues. After analysis (not saying I did it), we may find users trying to get Nightlies (and failing to build for various reasons). If so, we can think of prebuild code/archives. I mentioned Containers only to achieve the latter like extracting builds for different OS/distro flavors. I agree with Mathieu that containers are (still) not common enough or available to run on existing installations (with no admin access for ordinary users); let alone running MPI+containers(+GPU).

**Comment by oschuett** [comment]
> 1. build supported configurations of CP2K in a quick and reliable way (make, ninja)

Actually, our current Make-based system does a fine job here. It's reliable, fast, and portable. 

> 2. verify mandatory or optional requirements and dependencies (cmake, autotools, etc.) ...

Yes, that functionality should indeed be moved out of the Toolchain and into a separate configure script (#1095). While CMake could get the job done, a Python script would probably be shorter and easier to maintain.

>  From that point-of-view it would make more sense to provide a Spack environment configuration ... and provide a small wrapper around it.

The bulk of the work is packaging our dependencies - especially the esoteric ones like libvori or ROCm-Aware MPI. Replacing our Toolchain with Spack would simply mean forcing that packaging work onto our developers, which would certainly slow them down. Sacrificing our development speed is a high price to pay. Besides, our [Spack package](https://spack.readthedocs.io/en/latest/package_list.html?highlight=cp2k#cp2k) is actually already in pretty good shape.


So again, I think we should focus on improving CP2K itself. Staying relevant is the best way to incentivize distributions, vendors, and computing centers to invest more in CP2K.

**Reply by mtaillefumier** [reply]
> > 1. build supported configurations of CP2K in a quick and reliable way (make, ninja)
> 
> Actually, our current Make-based system does a fine job here. It's reliable, fast, and portable.

The actual build system makes the assumption that we are able to configure cp2k manually and properly which is not the case. Make was never designed for that purpose and will not be modified to support the configuration part of any software. Just adding hip or opencl for instance lead to many issues with building cp2k and dbcsr. Adding openapi support will also be an issue when the time comes (it might be closer than we think). So no, the current build system is not as reliable as we may think it is.
 
> > 2. verify mandatory or optional requirements and dependencies (cmake, autotools, etc.) ...
> 
> Yes, that functionality should indeed be moved out of the Toolchain and into a separate configure script (#1095). While CMake could get the job done, a Python script would probably be shorter and easier to maintain.

this should part of the configuration stage although cmake, autotools, gcc etc are most probably installed on very developer workstation and most certainly available on any HPC cluster. Having a script checking for that is not strictly needed. CMake and autotools can check the installed version and many other build systems have this functionality builtin as well. Dependency versioning should be part of the configuration stage not separated from it and most certainly not handled by hand.

> > From that point-of-view it would make more sense to provide a Spack environment configuration ... and provide a small wrapper around it.
> 
> The bulk of the work is packaging our dependencies - especially the esoteric ones like libvori or ROCm-Aware MPI. Replacing our Toolchain with Spack would simply mean forcing that packaging work onto our developers, which would certainly slow them down. Sacrificing our development speed is a high price to pay. Besides, our [Spack package](https://spack.readthedocs.io/en/latest/package_list.html?highlight=cp2k#cp2k) is actually already in pretty good shape.

CP2K should not be responsible for packaging the dependencies. Let package managers and libraries developers make this work as they know best how the library works. What do we gain by not using building recipe for cp2k dependencies that actually developers of these libraries wrote. 

As @dev-zero pointed out, the toolchain does the work of a package manager with the caveat that we have to maintain the tool, the recipe to build the dependencies etc... The argument that the burden of maintaining spack rules imply that we have to maintain spack rules for the dependencies is misleading as almost all dependencies already have spack recipe maintained by the developers of the said library.  The only thing we have to maintain with any decent package manager are the rules for building cp2k (not its dependencies). It is far less work than maintaining the complete stack. 

I will take an example why the toolchain will show its limits very quickly (if not already) and why it can not be scaled. SIRIUS has an optional dependency called nlcg. We currently have no script in the toolchain that compile it why ? nlcg depends on kokkos which it self depends on many others dependencies, etc... A problem arises from this thought. Where do we stop the process of building dependencies because we may have secondary dependencies that depend on other dependencies as well etc... Ultimately we would have to write a script for every single dependency and maintain these scripts ourselves while this work is already done by the developers of the unnamed dependency. 
 
> So again, I think we should focus on improving CP2K itself. Staying relevant is the best way to incentivize distributions, vendors, and computing centers to invest more in CP2K.

Computing centers invest a lot of time in cp2k already. Vendors have priorities that go far beyond the cp2k project and distributions will never directly contribute to cp2k code base as it is not their role. I think we all agree that we would like to spend more time on cp2k itself but right now we are limited by problems that can be solved by using tools that are already there. We must accept that we can not handle absolutely everything by hand.

**Comment by oschuett** [comment]
> SIRIUS has an optional dependency called nlcg. lcg depends on kokkos which it self depends on many others dependencies, etc... A problem arises from this thought. Where do we stop the process of building dependencies because we may have secondary dependencies that depend on other dependencies as well etc... 

Actually, looking at [Jürg's table](https://github.com/cp2k/cp2k/discussions/1989#discussioncomment-2356654) it becomes rather obvious that SIRIUS is special with regard to secondary dependencies. Personally, I'd say SIRIUS took a wrong turn and we should not encourage such architectures.

>  The argument that the burden of maintaining spack rules imply that we have to maintain spack rules for the dependencies is misleading as almost all dependencies already have spack recipe maintained by the developers of the said library.

Unfortunately, it's not that simple. Take e.g. my [lastest change](https://github.com/cp2k/cp2k/commit/75fc08f981b99c117ac3ce972113015b450ad285) to the Toolchain in which I upgraded ELPA to version 2021.11.002. If our CI were based on Spack I'd have to go through the following steps:

1. Since ELPA 2021.11.002 is not yet available in [Spack](https://github.com/spack/spack/blob/5a72204d38c9c0b5dc67df3ed2f455a3ddeb2727/var/spack/repos/builtin/packages/elpa/package.py), I have to open a pull request on the Spack repository for that. Currently, there are 600 open pull requests. The [last time](https://github.com/spack/spack/pull/28134) a new ELPA version was added it took 10 days for the PR to be merged.

2. Afterwards the responsible thing to do would be to wait for the next official Spack release, which take 1-4 months. Ok, I'm obviously going to skip this step and just upgrade to some recent developer version.

3. Upgrading Spack will not only pull in my change to the ELPA package, but also random changes to other packages. Since we're using a developer version of Spack there is a non-neglitable chance that one of those changes will break cp2k. In that case I have to investigate a totally unrelated issue and open another PR on the Spack repository.

4. Ok, let's assume I got lucky and nothing broke. Then I can now point cp2k's environment config at the new ELPA package. **Only at this last step does our CI run the cp2k regression and performance tests.** It is not uncommon for our tests to discover a problem that triggers a new upstream release. In that case the whole dance starts over at step 1.

**Reply by alazzaro** [reply]
> 
> > The argument that the burden of maintaining spack rules imply that we have to maintain spack rules for the dependencies is misleading as almost all dependencies already have spack recipe maintained by the developers of the said library.
> 
> Unfortunately, it's not that simple. Take e.g. my [lastest change](https://github.com/cp2k/cp2k/commit/75fc08f981b99c117ac3ce972113015b450ad285) to the Toolchain in which I upgraded ELPA to version 2021.11.002. If our CI were based on Spack I'd have to go through the following steps:
> 
> 1. Since ELPA 2021.11.002 is not yet available in [Spack](https://github.com/spack/spack/blob/5a72204d38c9c0b5dc67df3ed2f455a3ddeb2727/var/spack/repos/builtin/packages/elpa/package.py), I have to open a pull request on the Spack repository for that. Currently, there are 600 open pull requests. The [last time](https://github.com/spack/spack/pull/28134) a new ELPA version was added it took 10 days for the PR to be merged.
> 2. Afterwards the responsible thing to do would be to wait for the next official Spack release, which take 1-4 months. Ok, I'm obviously going to skip this step and just upgrade to some recent developer version.
> 3. Upgrading Spack will not only pull in my change to the ELPA package, but also random changes to other packages. Since we're using a developer version of Spack there is a non-neglitable chance that one of those changes will break cp2k. In that case I have to investigate a totally unrelated issue and open another PR on the Spack repository.
> 4. Ok, let's assume I got lucky and nothing broke. Then I can now point cp2k's environment config at the new ELPA package. **Only at this last step does our CI run the cp2k regression and performance tests.** It is not uncommon for our tests to discover a problem that triggers a new upstream release. In that case the whole dance starts over at step 1.

Maybe you are already expert of spack, so you know that already... In any case, for the case you describe, I advise to look at personal installation files with spack, for example:

https://doku.lrz.de/display/PUBLIC/Building+software+in+user+space+with+spack#Buildingsoftwareinuserspacewithspack-CreatingYourOwnPackage

In other words, you use spack as a tool for installing any code you like with dependencies (and all the other functionalities), people have documentation for that (so you don't need to describe fancy solutions), and I'm not sure you can do anything better with the toolchain...

**Comment by oschuett** [comment]
Thanks for the link @alazzaro! I was aware of that option, but I believe this is not what @dev-zero originally proposed. Yes, we could fork the Spack packages. However, then somebody would still have to maintain the upstream packages. For our users nothing would change.

> In other words, you use spack as a tool for installing ...  and I'm not sure you can do anything better with the toolchain...

Actually, I believe the opposite is true :-) Admittedly our Toolchain looks a bit spaghetti right now, but it's actually conceptually very simple. And I'm sure that after a bit of refactoring this will become visible again.

Spack, on the other hand, is extremely complex already on the conceptual level. Its [user manual](https://spack.readthedocs.io/_/downloads/en/v0.11.2/pdf/) exceeds 300 pages. All of this is needed because Spack is solving a very complicated problem. However, we would use it for a different, much simpler task. Presumably we'd use less than 1% of Spack's features. Hence, I believe we're better off with our boring home-grown solution.

In the end, there is simply no silver bullet. The amount of packaging work is a conserved quantity. We can only reduce it by removing dependencies. Everything else is just pushing the workload around.

**Reply by mtaillefumier** [reply]
> Spack, on the other hand, is extremely complex already on the conceptual level. Its [user manual](https://spack.readthedocs.io/_/downloads/en/v0.11.2/pdf/) exceeds 300 pages. All of this is needed because Spack is solving a very complicated problem. However, we would use it for a different, much simpler task. Presumably we'd use less than 1% of Spack's features. Hence, I believe we're better off with our boring home-grown solution.

I am not sure to comprehend the argument behind the documentation length of a tool. How do you correlate the difficulty to use a tool with the length of its manual. Anyone can use spack just by reading the first chapter (ok manybe the first two chapters). The GCC manual is probably as long or even longer and we still use it. MKL manual is more than a 1000 pages long and we use 0.5 % of all the functionalities. We also use mathematics despite the huge documentation / textbooks / articles, etc that are available. My point simply is; we do not need to know or read everything to be able to use a tool especially a package management tool. Often users just need to know the command to install a package not the details how to compile it etc...

**Reply by haampie** [reply]
> The amount of packaging work is a conserved quantity. We can only reduce it by removing dependencies. Everything else is just pushing the workload around.

If you (a) distribute this problem over a community and (b) worry about your direct dependencies only, you can reduce this load.

> Presumably we'd use less than 1% of Spack's features.

Note that these features include for instance deployment to HPC centers. You may not use them yourself, but they are used by others to make your software available.

**Comment by rathann** [comment]
Speaking as distribution packager (Fedora), I would very much prefer a standard build system like autotools, cmake or, well, anything that does let me build cp2k with Fedora standard compiler flags, mandatory and a (some) optional dependencies enabled, and ensures minimum required versions are installed. Right now, I'm patching one of the included `arch` files to make sure I'm following the latest changes. This breaks with every release and I have to rebase my patch, but the alternative would be to simply keep my own `arch` file which I would need to update with each release anyway.

Oh, and I'd like to be able to use system-installed `fypp` without patching Makefiles, too.

By the way, which is the best forum to discuss regression test suite failures in Fedora build? I'm getting one test failing consistently across all architectures as well as some others that are arch-specific.

**Reply by mtaillefumier** [reply]
I am also using Fedora and like you I experience continuous build failures when building cp2k. It pretty much breaks every time I want to add a new functionality to cp2k code bas or want to test a specific configuration. Forgetting to remove a `-DFEATURE` or missing `-Ifeature/include`  are often the origin of the problem.

Many colleagues working for both HPC centers and private companies share the same view point. A conventional build system is far better than the current system in place. I and few other are actively working on the cmake build but the combination of tools such as fypp (for the pre-processing) and the -D flags logic makes it a bit more challenging. Hopefully this will be sorted out and a PR will be created.

the github repository is actually the right place to discuss test-suite related issues.

**Comment by tgamblin** [comment]
I was pointed to this discussion and I thought I’d add a couple points:

1. We are aiming to offer binaries for a subset of Spack packages by June, and I think this could simplify the installation process for many.
2. A binary distribution ensures not only that some people do not have to build, but also that custom builds are more likely to work. We’ve had good luck with this for LLNL dev teams, whose open source packages would go out of sync with the Spack `develop` branch, but now are tested (along with their dependencies) every time a PR that could affect them is submitted.
3. I’d be happy to discuss how we might add some relevant configuration of `cp2k` to our CI.
4. If folks want faster merge times, we’d be happy to entrust someone in the `cp2k` team with merge privileges for it. We have a bot now that notifies maintainers every time their packages are in PRs, and that could also speed things up.

In Spack, we don’t have a strong opinion about which build system the project uses, so long as we can point it to arbitrary external dependency prefixes. That said, I like to point people to this talk:

https://youtu.be/sBP17HQAQjk

as I think “Don’t package your software, write packagable software” is a good rule of thumb.

**Reply by mtaillefumier** [reply]
It is really enlightening.

**Reply by mtaillefumier** [reply]
It is a bit extreme (optional dependency versus mandatory but I understand the point he is making). it really gives solutions to common problems they have (and actually developers have as well). The simpler and the most standard the build system is the easier it is for everyone to use it. This includes users, devs, admins and packages managers alike. This talk is really informative and goes towards the direction some of us have described earlier in this thread. Don't try to outsmart the build system is my favorite

**Reply by bcumming** [reply]
"Don't package your software, write packagable software" is a more succinct version of my comment below.

**Comment by oschuett** [comment]
Thanks for chiming in @tgamblin!

> If folks want faster merge times, we’d be happy to entrust someone in the cp2k team with merge privileges for it. 

Wow, I didn't see this one coming :-) Yes, having the ability to unblock our developers would indeed alleviate my biggest concern.

>  That said, I like to point people to this talk: https://youtu.be/sBP17HQAQjk

Thanks for sharing! For a C++ project I'd completely agree. However, CP2K is written in Fortran, which conventional tools rarely support well.

> A conventional build system is far better than the current system in place.

Maybe :-) The idea of switching to cmake has been floating around for almost 10 years. As you acknowledged yourself, it's not that easy.

> I and few other are actively working on the cmake build ... Hopefully this will be sorted out and a PR will be created.

Yes, this would be great. Can you give an ETA?

If this discussion has made one thing clear, it's the general need for a build configuration tool.
So, if the cmake migration doesn't pan out, then I'd like to take a stab at my proposal #1095.

**Reply by mtaillefumier** [reply]
> Thanks for chiming in @tgamblin!
> 
> > If folks want faster merge times, we’d be happy to entrust someone in the cp2k team with merge privileges for it.
> 
> Wow, I didn't see this one coming :-) Yes, having the ability to unblock our developers would indeed alleviate my biggest concern.
> 
> > That said, I like to point people to this talk: https://youtu.be/sBP17HQAQjk
> 
> Thanks for sharing! For a C++ project I'd completely agree. However, CP2K is written in Fortran, which conventional tools rarely support well.

The speaker was giving a talk at a cpp conference. Of course he would take examples that target c++. However the language is completely irrelevant here. The message is clear 

- custom build systems are a pain compared to standard one. 
- When using a standard build system, use builtin functions as much as possible, etc...
- package management should be done outside the library or app. 
 
> > A conventional build system is far better than the current system in place.
> 
> Maybe :-) The idea of switching to cmake has been floating around for almost 10 years. As you acknowledged yourself, it's not that easy.
> 
> > I and few other are actively working on the cmake build ... Hopefully this will be sorted out and a PR will be created.
> 
> Yes, this would be great. Can you give an ETA?

I have the basics working but I still need to test builds with hip and cuda (the second one works) as most of the dependencies already have cmake modules. The only real troubling case is the several flavors of blas and lapack. So far I only cover MKL, openblas and blas and lapack (from netlib). libsci will be next. Adding other variants requires writing a cmake module for the specific library. 

There is a cmake module for blas and lapack but it does not return the include directories which can be needed. 

Things are complicated because of the incoherent convention for the DFLAGS and the special treatment reserved to the fortran files. A typical example is -D__FEATURE_CUDA and -D__FEATURE_HIP. I would simply use -D__FEATURE_GPU and have a -D__CUDA or -D__HIP. 
 
> If this discussion has made one thing clear, it's the general need for a build configuration tool. So, if the cmake migration doesn't pan out, then I'd like to take a stab at my proposal #1095.

We still have to tinker with makefiles if we want to use oneapi, opencl, or different accelerators which means it does not solve the problem. NB : We already have this functionality with autotools. Result nobody use it which means it is outdated. It is not a solution in any cases.

**Reply by rathann** [reply]
> > > I and few other are actively working on the cmake build ... Hopefully this will be sorted out and a PR will be created.
> > 
> > Yes, this would be great. Can you give an ETA?
> 
> I have the basics working but I still need to test builds with hip and cuda (the second one works) as most of the dependencies already have cmake modules. The only real troubling case is the several flavors of blas and lapack. So far I only cover MKL, openblas and blas and lapack (from netlib). libsci will be next. Adding other variants requires writing a cmake module for the specific library.

[Fedora has switched](https://fedoraproject.org/wiki/Changes/FlexiBLAS_as_BLAS/LAPACK_manager) to [FlexiBLAS](https://www.mpi-magdeburg.mpg.de/projects/flexiblas) which is a thin wrapper around various BLAS+LAPACK implementations. Perhaps supporting only that and adding support for any missing implementations there would be the way forward? It'd certainly simplify the BLAS+LAPACK support in CP2K.

> There is a cmake module for blas and lapack but it does not return the include directories which can be needed.
> 
> Things are complicated because of the incoherent convention for the DFLAGS and the special treatment reserved to the fortran files. A typical example is -D__FEATURE_CUDA and -D__FEATURE_HIP. I would simply use -D__FEATURE_GPU and have a -D__CUDA or -D__HIP.
> 
> > If this discussion has made one thing clear, it's the general need for a build configuration tool. So, if the cmake migration doesn't pan out, then I'd like to take a stab at my proposal #1095.
> 
> We still have to tinker with makefiles if we want to use oneapi, opencl, or different accelerators which means it does not solve the problem. NB : We already have this functionality with autotools. Result nobody use it which means it is outdated. It is not a solution in any cases.

Autotools? I haven't noticed it before. It certainly wasn't mentioned in the main README. I would like to try enabling OpenCL in Fedora builds at some point though.

**Reply by hfp** [reply]
> I would like to try enabling OpenCL in Fedora builds at some point though.

@rathann Please create an issue if things are not ready by the time you want to go for OpenCL. I have missed to contribute CI-tests for CP2K and integration with the bespoken Toolchain. At this point, only the DBCSR component supports OpenCL but I will pickup other components going forward. An advantage of the DBCSR/OpenCL backend is decent performance without the need to supply tuned parameters (which are GPU-specific).

I think it is generally desirable (not only for the OpenCL bits) to be able to enable/disable GPU support at runtime rather than build-time.

**Reply by mtaillefumier** [reply]
@rathann Yes we have autotools but it is an incomplete build system and is outdated since I never took the effort to continue the support (guilty as charge). It will generate the arch file for sure but not all dependencies are included in the configure.ac script. 
I basically used it to generate the arch file instead of generating the Makefiles. I would have preferred the later option as it is the most standard way to use autotools but fate decided otherwise.

**Reply by mtaillefumier** [reply]
> [Fedora has switched](https://fedoraproject.org/wiki/Changes/FlexiBLAS_as_BLAS/LAPACK_manager) to [FlexiBLAS](https://www.mpi-magdeburg.mpg.de/projects/flexiblas) which is a thin wrapper around various BLAS+LAPACK implementations. Perhaps supporting only that and adding support for any missing implementations there would be the way forward? It'd certainly simplify the BLAS+LAPACK support in CP2K.

That's a very interesting point. Thanks for pointing that.

**Comment by bcumming** [comment]
I am chiming representing the perspictive of both developers, and those who have to build and provide applications on HPC systems for our users.                                                                                                                                          

I would propose focusing on clean separation of concerns between build configuration (e.g. CMake) and dependency management. In practice, this means that how you configure the build is independent of how dependencies are installed, it has zero knowledge about how the dependencies are installed.

Using CMake as an example (there are other choices), define the dependencies:

```cmake
find_package(nlcglib REQUIRED)
find_package(kokkos REQUIRED)
```

This requires that the dependency has been installed in a standard manner that can be found by CMake (which knows where to look for dependencies).

Then builders of the software  are free to provide those dependencies using whatever approach they choose:
* A developer can use a custom python script provided with the CP2K source code to install dependencies
* The staff at a HPC center responsible for providing CP2K on their cluster can integrate it into their spack configuration.
* Linux distribution packagers can use their package manager.
